pkg/notification/client.go

```diff
@@ -0,0 +1,159 @@
+package notification
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"text/template"
+
+	"github.com/opendevstack/pipeline/internal/kubernetes"
+	"github.com/opendevstack/pipeline/pkg/logging"
+	"github.com/opendevstack/pipeline/pkg/pipelinectxt"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	NotificationConfigMap   = "ods-notification"
+	UrlProperty             = "url"
+	MethodProperty          = "method"
+	ContentTypeProperty     = "contentType"
+	RequestTemplateProperty = "requestTemplate"
+	NotifyOnStatusProperty  = "notifyOnStatus"
+)
+
+type Client struct {
+	clientConfig     ClientConfig
+	httpClient       *http.Client
+	kubernetesClient kubernetes.ClientInterface
+}
+
+type ClientConfig struct {
+	Namespace string
+	Logger    logging.LeveledLoggerInterface
+}
+
+type PipelineRunResult struct {
+	PipelineRunURL string
+	OverallStatus  string
+	ODSContext     *pipelinectxt.ODSContext
+}
+
+type notificationConfig struct {
+	url            string
+	method         string
+	contentType    string
+	notifyOnStatus []string
+	template       *template.Template
+}
+
+func NewClient(config ClientConfig, kubernetesClient kubernetes.ClientInterface) (*Client, error) {
+	if config.Logger == nil {
+		config.Logger = &logging.LeveledLogger{Level: logging.LevelError}
+	}
+
+	return &Client{
+		clientConfig:     config,
+		httpClient:       &http.Client{},
+		kubernetesClient: kubernetesClient,
+	}, nil
+}
+
+func (c Client) readNotificationConfig(ctxt context.Context) (*notificationConfig, error) {
+	cm, err := c.kubernetesClient.GetConfigMap(ctxt, NotificationConfigMap, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to load %s ConfigMap: %v", NotificationConfigMap, err)
+	}
+
+	url, ok := cm.Data[UrlProperty]
+	if !ok {
+		return nil, fmt.Errorf("%s doesn't specify '%s' property", NotificationConfigMap, UrlProperty)
+	}
+
+	method, ok := cm.Data[MethodProperty]
+	if !ok {
+		return nil, fmt.Errorf("%s doesn't specify '%s' property", NotificationConfigMap, MethodProperty)
+	}
+
+	contentType, ok := cm.Data[ContentTypeProperty]
+	if !ok {
+		return nil, fmt.Errorf("%s doesn't specify '%s' property", NotificationConfigMap, ContentTypeProperty)
+	}
+
+	notifyOnStatus, ok := cm.Data[NotifyOnStatusProperty]
+	if !ok {
+		return nil, fmt.Errorf("%s doesn't specifiy '%s' property", NotificationConfigMap, NotifyOnStatusProperty)
+	}
+
+	decoder := json.NewDecoder(strings.NewReader(notifyOnStatus))
+	var notificationStatusValues []string
+	err = decoder.Decode(&notificationStatusValues)
+	if err != nil && err != io.EOF {
+		return nil, fmt.Errorf("decoding notification status properties failed: %w", err)
+	}
+
+	text, ok := cm.Data[RequestTemplateProperty]
+	if !ok {
+		return nil, fmt.Errorf("%s doesn't specify '%s' property", NotificationConfigMap, RequestTemplateProperty)
+	}
```

Comment:
I wonder if all this property reading could be done by mapping the map returned from the `ConfigMap` to `notificationConfig` directly. See https://stackoverflow.com/a/53289976.

Reply:
@henninggross and I were struggling with this one a bit: The code is definitely repetitive and we'd love to improve on it. However, using the `json.Marshall()` and `json.Unmarshall()` trick to map the `ConfigMap` map to a struct wouldn't work in our opinion since we still need to do some processing on the data to get it into the `NotificationConfig` struct (in particular transforming the template string to a proper `template.Template` and the JSON representation of the allowed status values to a string slice).